src/kernel/bitcoinkernel.h

```diff
@@ -0,0 +1,1271 @@
+    bool always_print_category_levels; //!< Prepend the log category and level to log messages.
+} kernel_LoggingOptions;
+
+/**
+ * A collection of status codes that may be issued by the script verify function.
+ */
+typedef enum {
+    kernel_SCRIPT_VERIFY_OK = 0,
+    kernel_SCRIPT_VERIFY_ERROR_TX_INPUT_INDEX, //!< The provided input index is out of range of the actual number of inputs of the transaction.
+    kernel_SCRIPT_VERIFY_ERROR_INVALID_FLAGS, //!< The provided bitfield for the flags was invalid.
+    kernel_SCRIPT_VERIFY_ERROR_INVALID_FLAGS_COMBINATION, //!< The flags very combined in an invalid way.
+    kernel_SCRIPT_VERIFY_ERROR_SPENT_OUTPUTS_REQUIRED, //!< The taproot flag was set, so valid spent_outputs have to be provided.
+    kernel_SCRIPT_VERIFY_ERROR_SPENT_OUTPUTS_MISMATCH, //!< The number of spent outputs does not match the number of inputs of the tx.
+} kernel_ScriptVerifyStatus;
+
+/**
+ * Script verification flags that may be composed with each other.
+ */
+typedef enum
+{
+    kernel_SCRIPT_FLAGS_VERIFY_NONE                = 0,
+    kernel_SCRIPT_FLAGS_VERIFY_P2SH                = (1U << 0), //!< evaluate P2SH (BIP16) subscripts
+    kernel_SCRIPT_FLAGS_VERIFY_DERSIG              = (1U << 2), //!< enforce strict DER (BIP66) compliance
+    kernel_SCRIPT_FLAGS_VERIFY_NULLDUMMY           = (1U << 4), //!< enforce NULLDUMMY (BIP147)
+    kernel_SCRIPT_FLAGS_VERIFY_CHECKLOCKTIMEVERIFY = (1U << 9), //!< enable CHECKLOCKTIMEVERIFY (BIP65)
+    kernel_SCRIPT_FLAGS_VERIFY_CHECKSEQUENCEVERIFY = (1U << 10), //!< enable CHECKSEQUENCEVERIFY (BIP112)
+    kernel_SCRIPT_FLAGS_VERIFY_WITNESS             = (1U << 11), //!< enable WITNESS (BIP141)
+
+    kernel_SCRIPT_FLAGS_VERIFY_TAPROOT             = (1U << 17), //!< enable TAPROOT (BIPs 341 & 342)
+    kernel_SCRIPT_FLAGS_VERIFY_ALL                 = kernel_SCRIPT_FLAGS_VERIFY_P2SH |
+                                                     kernel_SCRIPT_FLAGS_VERIFY_DERSIG |
+                                                     kernel_SCRIPT_FLAGS_VERIFY_NULLDUMMY |
+                                                     kernel_SCRIPT_FLAGS_VERIFY_CHECKLOCKTIMEVERIFY |
+                                                     kernel_SCRIPT_FLAGS_VERIFY_CHECKSEQUENCEVERIFY |
+                                                     kernel_SCRIPT_FLAGS_VERIFY_WITNESS |
+                                                     kernel_SCRIPT_FLAGS_VERIFY_TAPROOT
+} kernel_ScriptFlags;
+
+/**
+ * Chain type used for creating chain params.
+ */
+typedef enum {
+    kernel_CHAIN_TYPE_MAINNET = 0,
+    kernel_CHAIN_TYPE_TESTNET,
+    kernel_CHAIN_TYPE_TESTNET_4,
+    kernel_CHAIN_TYPE_SIGNET,
+    kernel_CHAIN_TYPE_REGTEST,
+} kernel_ChainType;
+
+/**
+ * A type-safe block identifier.
+ */
+typedef struct {
+    unsigned char hash[32];
+} kernel_BlockHash;
+
+/**
+ * Convenience struct for holding serialized data.
+ */
+typedef struct {
```

Comment:
i was wondering; these are trivially small structures, why don't we pass and return them by value instead of by pointer? this would avoid needing a special `kernel_byte_array_destroy` call to deallocate them